nx/lib/nx/binary_backend.ex

```diff
@@ -829,6 +829,30 @@ defmodule Nx.BinaryBackend do
   @impl true
   def conjugate(out, tensor), do: element_wise_unary_op(out, tensor, &Complex.conjugate/1)
 
+  @impl true
+  def real(out, %{type: {:c, s}} = tensor) do
+    data = to_binary(tensor)
+
+    result =
+      match_types [{:f, div(s, 2)}] do
+        for <<match!(real, 0), match!(_, 0) <- data>>, into: <<>>, do: real
```

Comment:
Can you try this?

```suggestion
    s = div(s, 2)

    result =
      for <<real::float-size(s), _::float-size(s) <- data>>, into: <<>>, do: real
```

The warning happens because the compiler knows that it can only be float.